man/man3/axis.3
.TH AXIS 3NEMO "
.SH NAME
xaxis, yaxis, axvar, xaxvar, yaxvar \- routines to draw various kinds of axes
.SH SYNOPSIS
.nf
\fB#include <axis.h>

void xaxis(x0, y0, l, tick, nticks, trans, label)
void yaxis(x0, y0, l, tick, nticks, trans, label)
real x0, y0, l, tick[];
int ntick;
axis_proc trans;
string label;

void axvar(sznum, szlab, scale)
real sznum, szlab, scale;

void xaxvar(ndig, tikup, tikdn, numdn, labdn)
void yaxvar(ndig, tikup, tikdn, numdn, labdn)
int ndig;
real tikup, tikdn, numdn, labdn;

extern real trans(/* real r */);
.fi
.SH DESCRIPTION
The \fIaxis(3NEMO)\fP routines provide a simple but general interface to 
the 
\fIyapp(3NEMO)\fP routines to draw and label X and Y-axes. Some helper
routines exist with which the axis control variables
(such as size and location of tick marks/values and labels - see below)
can be changed, though at startup they are all given reasonable values.
.PP
With the routines
\fIxaxis\fP and \fIyaxis\fP a X- and Y-axis can be drawn.
Axes can have
both tickmarks and labels (in which case tick values
are also drawn),
though the programmer can select any of them to be absent.
(\fPx0,y0\fP) is the lower/left edge of the x/y-axis, and \fIl\fP it's
length. \fInticks\fP controls the tick marks along the axis.
If positive, it is the length of 
the array \fItick\fP containing all of the
tick values; if negative, it's absolute value gives the number
of ticks (excluding the edges, where ticks are never drawn),
and \fItick[0],tick[1]\fP give the minimum and maximum value
along the axis. \fItrans\fP is
a (real-valued) function (of a single real-valued argument)
that converts world coordinates to
cm (the \fIyapp\fP units).  \fIlabel\fP is a text string,
that is used as a centered label along the axis. It can be the
NULL string, in which case no label and no tick values are plotted.
.PP
\fIaxvar\fP provides global control over both the X and Y-axis.
\fIsznum\fP is the size of the tick values (in cm).
\fIszlab\fP is the size of the text label (in cm).
\fIscale\fP rescales all (6) \fBreal\fP axis variables by multiplying it
with the old values. 
Any of these three parameters can be negative, in which case no action is
undertaken. Defaults are described below.
.PP
\fIxaxvar\fP and \fIyaxvar\fP provide control over the 
X and Y axis separately. 
\fIndig\fP controls the number of significant digits after the decimal
point, with which tick values are printed (though any trailing zeroes
are always deleted).
\fItikup\fP and \fItikdn\fP are the values of the length of the
tick up and down from the axis (the 'down' is usually 0.0, i.e.
ticks are normally not drawn outwards from an axis).
\fInumdn\fP is the distance from the axis where the tick values
are drawn.
\fIlabdn\fP is the distance from the axis where the label string
is drawn (default: 1.0cm for both X- and Y-axis).
Again, any variable can be given negative, in which case no action is
undertaken.
.SH BUGS
The real-valued 
axis control variables (\fI tikup, tikdn, numdn, sznum, labdn, szlab\fP for
both the X-axis and Y-axis)
are all in cm, with the usual \fIyapp(3NEMO)\fP convention that the
viewport is 20 by 20 cm. This confuses some english speaking nations.
.PP
There is no automatic tick value scaling.
.SH EXAMPLES
The following code will draw a linear X axis, from -4 to 4, 
with major and minor tickmarks; only the major tickmarks are labeled.
.nf

        real xrange[2] = {-4, 4};
        axis_proc xtrans();
        ..
        xaxvar(-1,0.2,-1,-1,-1);    \fI/* redefine big ticks */\fP
        xaxis(2, 2, 16, xrange, -7, xtrans, "My-X-label");
        xaxvar(-1,0.1,-1,-1,-1);    \fI/* set small ticks */\fP
        xaxis(2, 2, 16, xrange,-31, xtrans, NULL);
        xaxvar(-1,0.2,-1,-1,-1);    \fI/* and reset */\fP
        ..

        real xtrans(real x)
        {
            return 2 + 16*(x-xrange[0])/(xrange[1]-xrange[0]);
        }

.fi
.SH AXIS CONTROL VARIABLES
.nf
.ta +0.5i +0.5i +0.5i +1.0i
Variable	X	Y	Comments

ndig	2	2	number of digits after decimal point
tikup	0.2	0.2	tick-lenght upward from axis
tikdn	0.0	0.0	tick-lenght downward from axis
numdn	0.4	0.3	distance from axis where tick values drawn
sznum	0.32	0.32	size of tick values
labdn	1.0	1.0	distance from axis where axis label drawn
szlab	0.32	0.32	size of axis labels
.fi
.SH SEE ALSO
yapp(3NEMO)
.SH AUTHOR
Joshua Barnes, Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4i
xx-xxx-86	Code written          	JEB
25-apr-92	Document written	PJT
xx-xxx-xx	Added automatic tick value scaling option	PJT
16-feb-97	axis_proc for SINGLEPREC                  	pjt
.fi


src/kernel/misc/axis.c
/*
 * AXIS.C: routines to draw various kinds of axies.
 * 
 *	xx-xxx-86   created...					 JEB
 *	10-sep-90   trimval() declared as void in both places    PJT
 *	19-oct-90   index() -> strchr()                          PJT
 *	15-apr-95   ansi prototypes, no more ARGS		 PJT
 *      17-feb-97   proc prototype (for SINGLEPREC)		 pjt
 */

#include <stdinc.h>
#include <yapp.h>
#include <axis.h>

struct axisvars xaxisvar = { 2, 0.2, 0.0, 0.4, 0.32, 1.0, 0.32 };
struct axisvars yaxisvar = { 2, 0.2, 0.0, 0.3, 0.32, 1.0, 0.32 };

/*
 * Accessors for old-fashioned routines.
 */

#define nxdig  xaxisvar.ndig
#define xtikup xaxisvar.tikup
#define xtikdn xaxisvar.tikdn
#define xnumdn xaxisvar.numdn
#define xsznum xaxisvar.sznum
#define xlabdn xaxisvar.labdn
#define xszlab xaxisvar.szlab

#define nydig  yaxisvar.ndig
#define ytikrt yaxisvar.tikup
#define ytiklf yaxisvar.tikdn
#define ynumlf yaxisvar.numdn
#define ysznum yaxisvar.sznum
#define ylablf yaxisvar.labdn
#define yszlab yaxisvar.szlab

local void trimval(string);

/*
 * FORMALAXIS: if TRUE, draw axies in formal style.
 */

bool formalaxis;

/* #define FUDGEAXIS  TRUE */

/*
 * XAXIS: draw an x axis.
 */
 
void xaxis(
    real x0, real y0,	    /* axis starting point */
    real xl,		    /* axis length */
    real tick[],	    /* tick values or limits */
    int nticks,		    /* number of ticks */
    axis_proc xtrans,	    /* plotting transformation */
    string label	    /* label for axis */
)
{
    int i;
    real t, x;
    char val[32];

    plmove(x0, y0);
    plline(x0 + xl, y0);
    for (i = 0; i < abs(nticks); i++) {
	if (nticks > 0)
	    t = tick[i];
	else
	    t = tick[0] + (i + 1) * (tick[1] - tick[0]) / (1.0 - nticks);
	x = (*xtrans)(t);
	if (label != NULL) {
	    plmove(x, y0 + xtikup);
	    plline(x, y0 - xtikdn);
	    sprintf(val, "%-.*f", nxdig, t);
	    trimval(val);
	    pljust(0);
	    pltext(val, x, y0 - xnumdn, xsznum, 0.0);
	} else {
	    plmove(x, y0 - xtikup);
	    plline(x, y0 + xtikdn);
	}
    }
    if (label != NULL && *label != 0) {
	if (! formalaxis) {
	    pljust(0);
	    pltext(label, x0 + xl / 2, y0 - xlabdn, xszlab, 0.0);
	} else {
	    pljust(1);
#ifdef FUDGEAXIS
	    pltext(label, x0 + xl, y0 - xlabdn + xszlab/4.0, xszlab, 0.0);
#else
	    pltext(label, x0 + xl, y0 - xlabdn, xszlab, 0.0);
#endif
	}
    }
    pljust(-1);
}

/*
 * YAXIS: draw a y axis.
 */

void yaxis(
    real x0,
    real y0,    		/* axis starting point */
    real yl,		        /* axis length */
    real tick[],		/* tick values or limits */
    int nticks,		        /* number of ticks */
    axis_proc ytrans,		/* plotting transformation */
    string label		/* label for axis */
)
{
    int i;
    real t, y;
    char val[32];

    plmove(x0, y0);
    plline(x0, y0 + yl);
    for (i = 0; i < abs(nticks); i++) {
	if (nticks > 0)
	    t = tick[i];
	else
	    t = tick[0] + (i + 1) * (tick[1] - tick[0]) / (1.0 - nticks);
	y = (*ytrans)(t);
	if (label != NULL) {
	    plmove(x0 + ytikrt, y);
	    plline(x0 - ytiklf, y);
	    sprintf(val, "%-.*f", nydig, t);
	    trimval(val);
	    if (! formalaxis) {
		pljust(0);
		pltext(val, x0 - ynumlf, y, ysznum, 90.0);
	    } else {
		pljust(1);
		pltext(val, x0 - ynumlf, y, ysznum, 0.0);
	    }
	} else {
	    plmove(x0 - ytikrt, y);
	    plline(x0 + ytiklf, y);
	}
    }
    if (label != NULL && *label != 0) {
	if (! formalaxis) {
	    pljust(0);
	    pltext(label, x0 - ylablf, y0 + yl / 2, yszlab, 90.0);
	} else {
	    pljust(1);
#ifdef FUDGEAXIS
	    pltext(label, x0 - ylablf, y0 + yl - yszlab/4.0, yszlab, 0.0);
#else
	    pltext(label, x0 - ylablf, y0 + yl - yszlab/2.0, yszlab, 0.0);
#endif
	}
    }
    pljust(-1);
}

local void trimval(string val)
{
    char *ep;

    if (strchr(val, '.') != NULL) {
	ep = val + strlen(val);
	while (*--ep == '0')
	    *ep = 0;
	if (*ep == '.')
	    *ep = 0;
    }
}

/*
 * AXVAR: specify general parameters for axis subroutines.
 */

void axvar(real sn, real sl, real scale)
{
    if (sn > 0.0)
	xsznum = ysznum = sn;
    if (sl > 0.0)
	xszlab = yszlab = sl;
    if (scale > 0.0) {
	xtikup *= scale;
	xtikdn *= scale;
	xnumdn *= scale;
	xsznum *= scale;
	xlabdn *= scale;
	xszlab *= scale;
	ytikrt *= scale;
	ytiklf *= scale;
	ynumlf *= scale;
	ysznum *= scale;
	ylablf *= scale;
	yszlab *= scale;
    }
}

/*
 * XAXVAR: specify parameters for xaxis.
 */

void xaxvar(int nxd, real xtu, real xtd, real xnd, real xld)
{
    if (nxd >= 0)
	nxdig = nxd;
    if (xtu >= 0.0)
	xtikup = xtu;
    if (xtd >= 0.0)
	xtikdn = xtd;
    if (xnd >= 0.0) 
	xnumdn = xnd;
    if (xld >= 0.0)
	xlabdn = xld;
}

/*
 * YAXVAR: specify parameters for yaxis.
 * Imports: as for xaxvar.
 */

void yaxvar(int nyd, real ytr, real ytl, real ynl, real yll)
{
    if (nyd >= 0)
	nydig = nyd;
    if (ytr >= 0.0)
	ytikrt = ytr;
    if (ytl >= 0.0)
	ytiklf = ytl;
    if (ynl >= 0.0)
	ynumlf = ynl;
    if (yll >= 0.0)
	ylablf = yll;
}

#ifdef TESTBED

#include <getparam.h>

string defv[] = {
    "formalaxis=false",
    "xlabdn=1.0",
    "xszlab=0.32",
    "ylablf=1.0",
    "yszlab=0.32",
    NULL,
};

real xrange[] = { -2.0, 2.0 }, yrange[] = { 0.0, 4.0 };

nemo_main()
{
    real xtrans(real), ytrans(real);

    formalaxis = getbparam("formalaxis");
    xlabdn = getdparam("xlabdn");
    xszlab = getdparam("xszlab");
    ylablf = getdparam("ylablf");
    yszlab = getdparam("yszlab");
    plinit("***", 0.0, 20.0, 0.0, 20.0);
    xaxis( 2.0,  2.0, 16.0, xrange, -3, xtrans, "x");
    xaxis( 2.0, 18.0, 16.0, xrange, -3, xtrans, NULL);
    yaxis( 2.0,  2.0, 16.0, yrange, -7, ytrans, "y");
    yaxis(18.0,  2.0, 16.0, yrange, -7, ytrans, NULL);
    plstop();
}

real xtrans(real x)
{
    return (10.0 + 4.0 * x);
}

real ytrans(real y)
{
    return (2.0 + 4.0 * y);
}

#endif
